sklearn/model_selection/_search.py

```diff
@@ -234,63 +243,112 @@ class ParameterSampler(object):
+    >>> import numpy as np
+    >>> np.random.seed(0)
+    >>> param_grid = [{'a':[1, 2], 'b': expon()}, {'c':[3, 4], 'd': expon()}]
+    >>> param_list = list(ParameterSampler(param_grid, n_iter=[4,6]))
     >>> rounded_list = [dict((k, round(v, 6)) for (k, v) in d.items())
     ...                 for d in param_list]
-    >>> rounded_list == [{'b': 0.89856, 'a': 1},
-    ...                  {'b': 0.923223, 'a': 1},
-    ...                  {'b': 1.878964, 'a': 2},
-    ...                  {'b': 1.038159, 'a': 2}]
+    >>> rounded_list == [{'a': 1, 'b': 0.89856},
+    ...                  {'a': 1, 'b': 0.923223},
+    ...                  {'a': 2, 'b': 1.878964},
+    ...                  {'a': 2, 'b': 1.038159},
+    ...                  {'c': 3, 'd': 0.353159},
+    ...                  {'c': 3, 'd': 3.314912},
+    ...                  {'c': 3, 'd': 0.649446},
+    ...                  {'c': 4, 'd': 0.752674},
+    ...                  {'c': 3, 'd': 0.498872},
+    ...                  {'c': 4, 'd': 0.073685}]
     True
     """
     def __init__(self, param_distributions, n_iter, random_state=None):
-        self.param_distributions = param_distributions
         self.n_iter = n_iter
         self.random_state = random_state
 
-    def __iter__(self):
-        # check if all distributions are given as lists
-        # in this case we want to sample without replacement
-        all_lists = np.all([not hasattr(v, "rvs")
-                            for v in self.param_distributions.values()])
-        rnd = check_random_state(self.random_state)
-
-        if all_lists:
-            # look up sampled parameter settings in parameter grid
-            param_grid = ParameterGrid(self.param_distributions)
-            grid_size = len(param_grid)
-            n_iter = self.n_iter
+        if isinstance(param_distributions, Mapping):
+            # wrap dictionary in a singleton list to support either dict
+            # or list of dicts
+            param_distributions = [param_distributions]
+
+        # check if all entries are dictionaries
+        for dist in param_distributions:
+            if not isinstance(dist, dict):
+                raise TypeError('Parameter distributions is not a '
+                                'dict ({!r})'.format(dist))
 
-            if grid_size < n_iter:
-                warnings.warn(
-                    'The total space of parameters %d is smaller '
-                    'than n_iter=%d. Running %d iterations. For exhaustive '
-                    'searches, use GridSearchCV.'
-                    % (grid_size, self.n_iter, grid_size), UserWarning)
-                n_iter = grid_size
-            for i in sample_without_replacement(grid_size, n_iter,
-                                                random_state=rnd):
-                yield param_grid[i]
+        self.param_distributions = param_distributions
 
+    def __iter__(self):
+        param_distributions = self.param_distributions
```

Comment:
Why do you duplicate `param_distributions`?

Reply:
There is no need to duplicate.
```suggestion
```